### PR TITLE
configurable loss functions + SSIM, en/decoder filename now include trainer name

### PR DESCRIPTION
**Pull request by IUsedToBeAPygmy**
add configurable loss functions including SSIM (Structural Image Similarity)
written filenames now include trainer name

### CODE REVIEW

**Comment by Dev-iL** [comment]
Adding a link to [the relevant issue](https://github.com/deepfakes/faceswap/issues/142)...

**Comment by IUsedToBeAPygmy** [comment]
@gdunstone please review, I don't see any option to add/assign you as reviewer

**Comment by gdunstone** [comment]
I'll have a review once I'm back at my pc. (havent forgotten about ya)

File: scripts/train.py [hunk 1]
```diff
@@ -79,6 +79,16 @@ def parse_arguments(self, description, subparser, command):
                             type=int,
                             default=64,
                             help="Batch size, as a power of 2 (64, 128, 256, etc)")
+        parser.add_argument('-l', '--loss-function',
```
Comment (gdunstone) [inline scripts/train.py]: should the selection of loss function change the default  loss_mix?

File: plugins/LossFunction_l1.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
+from keras import losses
+
+# L1 loss function = Mean of Absolute Error
+
```
Comment (gdunstone) [inline plugins/LossFunction_l1.py]: Is there any benefit here to having this over just assigning LossFunction to the loss function signature from keras?

LossFunction = losses.mean_absolute_error

File: lib/ModelAE.py [hunk 3]
```diff
@@ -3,28 +3,36 @@
...
         self.decoder_A = self.Decoder()
         self.decoder_B = self.Decoder()
-
+        self.loss_function = mean_absolute_error
         self.initModel()
+        
+    #
+    # set_loss_function has to be called *before* the model is initialized!
+    #
+    def set_loss_function(self, loss_function):
```
Comment (gdunstone) [inline lib/ModelAE.py]: whats the reasoning behind this function over passing the loss function to __init__?

File: scripts/train.py (hunk 1 above)
Comment (IUsedToBeAPygmy) [reply scripts/train.py]: Currently it doesn't need to. The loss mix is currently only used with (both) the SSIM loss functions, and is set by default to a very low value hardly influencing results - experimentation with this value is needed to see whether SSIM actually improves results or is completely unnecessary.
This was purely added to make experimentation with the Keras.Contrib SSIM implementation possible - I have not had the time to actually test the results.

File: plugins/LossFunction_l1.py (hunk 2 above)
Comment (IUsedToBeAPygmy) [reply plugins/LossFunction_l1.py]: Wow. After reviewing this code I found it actually doesn't work - it will always use the default loss function as it's not being overwritten in time.

Hold on while I try to figure out how to undo this pull request and fix my code :)
